tools/wake-format/emitter.cpp

```diff
@@ -201,23 +201,54 @@ static size_t count_trailing_newlines(const token_traits_map_t& traits, const CS
   return it->second.after_bound.size();
 }
 
-static size_t count_allowed_newlines(const token_traits_map_t& traits, const CSTElement& node) {
-  FMT_ASSERT(node.isNode(), node,
-             "Expected node, Saw <" + std::string(symbolName(node.id())) + ">");
-  return count_leading_newlines(traits, node) + count_trailing_newlines(traits, node);
+// Determines if a given node would emit a leading comment if emitted.
+static bool has_leading_comment(const CSTElement& node, const token_traits_map_t& traits) {
+  return count_leading_newlines(traits, node) > 0;
+}
+
+// Determines if a given node would emit a trailing comment if emitted.
+static bool has_trailing_comment(const CSTElement& node, const token_traits_map_t& traits) {
+  return count_trailing_newlines(traits, node) > 0;
+}
+
+// Determizes of a doc is "flat" as a human would judge it. Human's don't
+// consider leading/trailing comments as invalidating flatness even though they
+// necessite extra newlines. Thus leading/trailing newlines are allowed.
+static bool is_human_flat(const wcl::doc& doc, const CSTElement& node,
```

Comment:
I think part of my issue with this is that there are actually 3 kinds of "flat", not just 2.

1) Actully flat meaning the docs has no (or 1 depending on how you define it) newlines
2) This definition which is good for defs/requires inside of blocks, or expressions inside of exploded operator lists...e.g. anything where we have a sequence we want to output vertically.
3) One where only trailing comments are not considered to ruin flatness but a leading comment would actully be cause to indent something.

For instance

The following should not be indented
```
require Some x = y
else failWithError "blarg" # fail fast
```

but this should be

The following should not be indented (in fact it has to be in order to be correct)
```
require Some x = y
else
    # fail fast
    failWithError "blarg"
```

That said the expression in the else case would be considered "human flat" as you described it. I think we need to refine our vocubulary here to consider all 3 cases and use them appropriately.

Reply:
How about this

1) strict_flat (doc has no newlines)
2) vertically_flat (what you currently call human flat. this name hints that we want to use this notion of flat in vertical output cases like blocks or exploded expressions)
3) weak_flat (the 3rd, as yet unnamed notion of "flat")

Reply:
Okay! This should be done! PTAL